README.md
# FORKED REPOSITORY

This repository was forked from github.com/mckaywrigley/chatbot-ui.git . The code was scanned by Snyk (no major vulnerabilities) - and, later, [with OpenAI](./testme.ipynb) - and we didn't see data flowing anywhere but OpenAI API).
It's forked and "frozen" to minimize the risk. In future we will, hopefully, have someone responsible for this project - but as of now we want to provide the service ASAP.


# Chatbot UI

Chatbot UI is an open source chat UI for AI models.

See a [demo](https://twitter.com/mckaywrigley/status/1640380021423603713?s=46&t=AowqkodyK6B4JccSOxSPew).

![Chatbot UI](./public/screenshots/screenshot-0402023.jpg)

## Updates

Chatbot UI will be updated over time.

Expect frequent improvements.

**Next up:**

- [ ] Sharing
- [ ] "Bots"

## Deploy

**Vercel**

Host your own live version of Chatbot UI with Vercel.

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/clone?repository-url=https%3A%2F%2Fgithub.com%2Fmckaywrigley%2Fchatbot-ui)

**Docker**

Build locally:

```shell
docker build -t chatgpt-ui .
docker run -e OPENAI_API_KEY=xxxxxxxx -e AUTH_AZURE_AD_CLIENT_ID=xxxxxxxx -e AUTH_AZURE_AD_TENANT_ID=xxxxxxxx -e AUTH_AZURE_AD_SECRET=xxxxxxxx -e NEXTAUTH_SECRET=xxxxxxxx -p 3000:3000 chatgpt-ui
```

Pull from ghcr:

```
docker run -e OPENAI_API_KEY=xxxxxxxx -p 3000:3000 ghcr.io/mckaywrigley/chatbot-ui:main
```

## Running Locally

**1. Clone Repo**

```bash
git clone https://github.com/mckaywrigley/chatbot-ui.git
```

**2. Install Dependencies**

```bash
npm i
```

**3. Provide OpenAI API Key**

Create a .env.local file in the root of the repo with your OpenAI API Key:

```bash
OPENAI_API_KEY=YOUR_KEY
```

> You can set `OPENAI_API_HOST` where access to the official OpenAI host is restricted or unavailable, allowing users to configure an alternative host for their specific needs.

> Additionally, if you have multiple OpenAI Organizations, you can set `OPENAI_ORGANIZATION` to specify one.

**4. Configure SSO for Auth0**

Add variables .env.local:
```bash
AUTH_AUTH0_CLIENT_ID="QXHosbmuKaegHNn9MMSAKUhRNb6JXvNp"
AUTH_AUTH0_SECRET="N-uPvuwXXluEkIT69E2RvEYxA7lHoUh4zMtjPzlGlJDCNdPzJhMTBjknGcSY9WGl"
AUTH_AUTH0_HOST="https://chatbot-ui-staging.eu.auth0.com"
NEXTAUTH_SECRET=xxxxx
```

where `NEXTAUTH_SECRET` is random string (`openssl rand -base64 32`)

**5. Run App**

```bash
npm run dev
```

**6. Use It**

You should be able to start chatting.

## Configuration

When deploying the application, the following environment variables can be set:

| Environment Variable              | Default value                  | Description                                                                                                                               |
| --------------------------------- | ------------------------------ | ----------------------------------------------------------------------------------------------------------------------------------------- |
| AUTH_AZURE_AD_CLIENT_ID           |                                | Client ID from Azure AD
| AUTH_AZURE_AD_TENANT_ID           |                                | Tenant ID from Azure AD
| AUTH_AZURE_AD_SECRET              |                                | Client Secret from Azure AD
| NEXTAUTH_SECRET                   |                                | Random crypto-string
| OPENAI_API_KEY                    |                                | The default API key used for authentication with OpenAI                                                                                   |
| OPENAI_API_HOST                   | `https://api.openai.com`       | The base url, for Azure use `https://<endpoint>.openai.azure.com`                                                                         |
| OPENAI_API_TYPE                   | `openai`                       | The API type, options are `openai` or `azure`                                                                                             |
| OPENAI_API_VERSION                | `2023-03-15-preview`           | Only applicable for Azure OpenAI                                                                                                          |
| OPENAI_ORGANIZATION               |                                | Your OpenAI organization ID                                                                                                               |
| DEFAULT_MODEL | `gpt-3.5-turbo` _(OpenAI)_ `gpt-35-turbo` _(Azure)_ | The default model to use on new conversations |
| NEXT_PUBLIC_DEFAULT_TEMPERATURE   | 1                              | The default temperature to use on new conversations                                                                                       |
| GOOGLE_API_KEY                    |                                | See [Custom Search JSON API documentation][GCSE]                                                                                          |
| GOOGLE_CSE_ID                     |                                | See [Custom Search JSON API documentation][GCSE]                                                                                          |

If you do not provide an OpenAI API key with `OPENAI_API_KEY`, users will have to provide their own key.

If you don't have an OpenAI API key, you can get one [here](https://platform.openai.com/account/api-keys).

## Contact

If you have any questions, feel free to reach out to Mckay on [Twitter](https://twitter.com/mckaywrigley).

[GCSE]: https://developers.google.com/custom-search/v1/overview


CHANGELOG.md
## [1.1.1](https://git.epam.com/epm-gpt/common/chatbot-ui/compare/1.1.0...1.1.1) (2023-05-31)

## [1.1.0](https://git.epam.com/epm-gpt/common/chatbot-ui/compare/1.0.0...1.1.0) (2023-05-30)

### Bug Fixes

- helm deployment fix [no-build] ([fa3c728](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/fa3c728e24d5a2c1afd5b725213be46f4eaaea97))
- show errors messages according to server response ([b08c0ff](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/b08c0ff2071862afbe6c46826f56b1fd43a7ddc7))
- url fix in helm [no-build] ([d6f6eaf](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/d6f6eaf71dfd116ba2095410a7386a618319dba0))

### Features

- export single conversation ([d29e776](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/d29e7764866c010c11bc95047e955b7a42baf18d))

## [1.0.0](https://git.epam.com/epm-gpt/common/chatbot-ui/compare/...1.0.0) (2023-04-28)

### Bug Fixes

- change message edit logic ([#144](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/144)) ([4fbb5e1](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/4fbb5e1f79c9d279238b0adb24ee4c4063c275f7))
- chat messages disappear on edit ([#575](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/575)) ([836c246](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/836c24680b10e3014ca30c49f5105f796e830864)), closes [#574](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/574)
- disable button if messageContent is empty ([#137](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/137)) ([d7fdcd0](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/d7fdcd0dfe5df2e3b9d5d8735920d5d13bdfc901))
- fix fonts ([#194](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/194)) ([c0b1b2e](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/c0b1b2eadba4f2b08dfdb5daea17a498dc46b338))
- In light mode, the arrow in the "scroll to bottom" icon is difficult to see [#299](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/299) ([#304](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/304)) ([4d7adf4](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/4d7adf477a423c2f9ed52115a2615ba20457ccd1))
- issues/498 ([#499](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/499)) ([6d53ee7](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/6d53ee7d12fb7edc8fa8dc6f0773bbc59e1999ad))
- label layout shift ([#338](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/338)) ([aefec17](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/aefec17525b664ba3e2d82512fa0196622925182))
- layout shifting on small devices ([#258](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/258)) ([abdcd45](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/abdcd4508da787605119af946b0a1f5b40d8b965))
- missing translation ([#206](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/206)) ([bf8830e](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/bf8830e1a59ff89c1d7599449b8f62b9925599e7))
- modal scrolling on small devices ([#527](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/527)) ([f61e91d](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/f61e91d07f240c227aaff642ffb4127b64fa460f))
- optimize mobile layout ([#606](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/606)) ([dcb2810](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/dcb2810d4692d14d7995672f42f7673d4e0abb66))
- overlap between plus and prompt menu icon ([#230](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/230)) ([5aa5be3](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/5aa5be3f435981b053ee8cab7205b6107ff34dc5))
- prompt sidebar not showing on small devices ([#232](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/232)) ([5d31947](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/5d31947ab9392a66378859576ee12720b793f8f5))
- Promptbar initializes in open state for Mobile ([#542](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/542)) ([7c259b9](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/7c259b98cbad3624ca75895fddbf50c674d06ab1))
- prompts list height ([#340](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/340)) ([3650d8d](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/3650d8d7bfa5debb0167c3512defe3fced87a393))
- properly trigger prompt modal close ([#309](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/309)) ([2114b72](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/2114b7296e5f2d3099ed997a8eda1265c9e4841e))
- resolve Enter event conflict during CJK IME ([#253](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/253)) ([a73ef2b](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/a73ef2b8cf1dad6d4f1e78cb3b28f9d4e228c3f2))
- small ui issue ([#311](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/311)) ([5cdd3e5](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/5cdd3e56b7c25a557d4afca3b8c241ef92fec38d))
- some ui issues ([#346](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/346)) ([91cbe0b](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/91cbe0b104e8e43f724446e56c0b86dd8a1173df))
- sroll down button overlaping chat input ([#564](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/564)) ([00f6189](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/00f6189a42a7878d8b6aa7ae35e70d173ac180a5))
- support japanese typing ([3ba74d9](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/3ba74d99fbcc965092b5a2bc898b2216383fc6a4))
- temperature font color ([#556](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/556)) ([51fa461](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/51fa461965cede28788569a9310cd38254d9db7d))
- translation omission ([#281](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/281)) ([6ef83b0](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/6ef83b0cb6b56667030e8f23327b8462d1de1b96))
- update es translations ([#167](https://git.epam.com/epm-gpt/common/chatbot-ui/issues/167)) ([e182762](https://git.epam.com/epm-gpt/common/chatbot-ui/commit/e18276223b38e8e6c4b35ec9ce7456cb5e52734d))

types/openai.ts
export type OpenAIEntityModelType = 'model';
export type OpenAIEntityApplicationType = 'application';
export type OpenAIEntityAssistantType = 'assistant';
export type OpenAIEntityAddonType = 'addon';

export interface ProxyOpenAIEntity<T = OpenAIEntityModelType> {
  id: string;
  object: T;
  capabilities?: {
    embeddings: boolean;
    chat_completion: boolean;
  };
  addons?: string[];
}

export interface OpenAIEntity {
  id: string;
  name: string;
  type:
    | OpenAIEntityModelType
    | OpenAIEntityApplicationType
    | OpenAIEntityAssistantType
    | OpenAIEntityAddonType;
  selectedAddons?: string[];
}

export interface OpenAIEntityModel extends Omit<OpenAIEntity, 'type'> {
  maxLength: number; // maximum length of a message
  requestLimit: number;
  isDefault?: boolean;
  type:
    | OpenAIEntityModelType
    | OpenAIEntityApplicationType
    | OpenAIEntityAssistantType;
}

export interface OpenAIEntityAddon extends Omit<OpenAIEntity, 'type'> {
  type: 'addon';
}

export enum OpenAIEntityModelID {
  GPT_3_5 = 'gpt-3.5-turbo',
  GPT_3_5_AZ = 'gpt-35-turbo',
  GPT_4 = 'gpt-4',
  GPT_4_32K = 'gpt-4-32k',
  BISON_001 = 'chat-bison@001',
  AMAZON_TITAN_TG1_LARGE = 'amazon.titan-tg1-large',
  AI21_J2_GRANDE_INSTRUCT = 'ai21.j2-grande-instruct',
  AI21_J2_JUMBO_INSTRUCT = 'ai21.j2-jumbo-instruct',
  ANTHROPIC_CLAUDE_INSTANT_V1 = 'anthropic.claude-instant-v1',
  ANTHROPIC_CLAUDE_V1 = 'anthropic.claude-v1',
  STABILITY_STABLE_DIFFUSION_XL = 'stability.stable-diffusion-xl',
  GPT_WORLD = 'gpt-world',
  MIRROR = 'mirror',
  EPAM10K = 'epam10k',
  ASSISTANT10K = 'assistant-10k',
  EPAM10K_SEMANTIC_SEARCH = 'epam10k-semantic-search',
  EPAM10K_GOLDEN_QNA = 'epam10k-golden-qna',
}

export enum OpenAIEntityAddonID {
  ADDON_EPAM10K_SEMANTIC_SEARCH = 'addon-epam10k-semantic-search',
  ADDON_EPAM10K_GOLDEN_QNA = 'addon-epam10k-golden-qna',
  ADDON_WOLFRAM = 'addon-wolfram',
}

export const OpenAIEntityAddons: Record<
  OpenAIEntityAddonID,
  OpenAIEntityAddon
> = {
  [OpenAIEntityAddonID.ADDON_EPAM10K_SEMANTIC_SEARCH]: {
    id: OpenAIEntityAddonID.ADDON_EPAM10K_SEMANTIC_SEARCH,
    name: 'EPAM10K Semantic Search',
    type: 'addon',
  },
  [OpenAIEntityAddonID.ADDON_EPAM10K_GOLDEN_QNA]: {
    id: OpenAIEntityAddonID.ADDON_EPAM10K_GOLDEN_QNA,
    name: 'EPAM10K Golden QNA',
    type: 'addon',
  },
  [OpenAIEntityAddonID.ADDON_WOLFRAM]: {
    id: OpenAIEntityAddonID.ADDON_WOLFRAM,
    name: 'Wolfram',
    type: 'addon',
  },
};

// in case the `DEFAULT_MODEL` environment variable is not set or set to an unsupported model
export const fallbackModelID = OpenAIEntityModelID.GPT_3_5_AZ;

export const defaultModelLimits = {
  maxLength: 24000,
  requestLimit: 6000,
};

export const OpenAIEntityModels: Record<string, OpenAIEntityModel> = {
  [OpenAIEntityModelID.GPT_3_5]: {
    id: OpenAIEntityModelID.GPT_3_5,
    name: 'GPT-3.5',
    maxLength: 12000,
    requestLimit: 3000,
    type: 'model',
  },
  [OpenAIEntityModelID.GPT_3_5_AZ]: {
    id: OpenAIEntityModelID.GPT_3_5_AZ,
    name: 'GPT-3.5',
    maxLength: 12000,
    requestLimit: 3000,
    type: 'model',
  },
  [OpenAIEntityModelID.GPT_4]: {
    id: OpenAIEntityModelID.GPT_4,
    name: 'GPT-4',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.GPT_4_32K]: {
    id: OpenAIEntityModelID.GPT_4_32K,
    name: 'GPT-4-32K',
    maxLength: 96000,
    requestLimit: 24000,
    type: 'model',
  },
  [OpenAIEntityModelID.BISON_001]: {
    id: OpenAIEntityModelID.BISON_001,
    name: 'PaLM2 (bison)',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.AMAZON_TITAN_TG1_LARGE]: {
    id: OpenAIEntityModelID.AMAZON_TITAN_TG1_LARGE,
    name: 'AWS (Titan)',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.AI21_J2_GRANDE_INSTRUCT]: {
    id: OpenAIEntityModelID.AI21_J2_GRANDE_INSTRUCT,
    name: 'AI21 (Jurassic-2 Grande)',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.AI21_J2_JUMBO_INSTRUCT]: {
    id: OpenAIEntityModelID.AI21_J2_JUMBO_INSTRUCT,
    name: 'AI21 (Jurassic-2 Jumbo)',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.ANTHROPIC_CLAUDE_INSTANT_V1]: {
    id: OpenAIEntityModelID.ANTHROPIC_CLAUDE_INSTANT_V1,
    name: 'Anthropic (Claude Instant)',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.ANTHROPIC_CLAUDE_V1]: {
    id: OpenAIEntityModelID.ANTHROPIC_CLAUDE_V1,
    name: 'Anthropic (Claude)',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.STABILITY_STABLE_DIFFUSION_XL]: {
    id: OpenAIEntityModelID.STABILITY_STABLE_DIFFUSION_XL,
    name: 'Stable Diffusion XL',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.EPAM10K_GOLDEN_QNA]: {
    id: OpenAIEntityModelID.EPAM10K_GOLDEN_QNA,
    name: 'EPAM10K Golden QNA',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.EPAM10K_SEMANTIC_SEARCH]: {
    id: OpenAIEntityModelID.EPAM10K_SEMANTIC_SEARCH,
    name: 'EPAM10K Semantic Search',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'model',
  },
  [OpenAIEntityModelID.GPT_WORLD]: {
    id: OpenAIEntityModelID.GPT_WORLD,
    name: 'GPT World',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'application',
  },
  [OpenAIEntityModelID.MIRROR]: {
    id: OpenAIEntityModelID.MIRROR,
    name: 'Echo',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'application',
  },
  [OpenAIEntityModelID.EPAM10K]: {
    id: OpenAIEntityModelID.EPAM10K,
    name: 'EPAM10K',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'application',
  },
  [OpenAIEntityModelID.ASSISTANT10K]: {
    id: OpenAIEntityModelID.ASSISTANT10K,
    name: 'ASSISTANT10K',
    maxLength: 24000,
    requestLimit: 6000,
    type: 'assistant',
    selectedAddons: [OpenAIEntityAddonID.ADDON_EPAM10K_SEMANTIC_SEARCH],
  },
};
